### ISSUE TITLE
Feature request: Remove unnecessary parenthesis

### ISSUE THREAD

**Issue by jan-jezek** [SAP/abap-cleaner#105]
Hello,

Would it be possible to create a rule, that removes unnecessary parenthesis as described in the abaplint rule: https://rules.abaplint.org/many_parentheses/

Example:
``` abap
IF ( variable IS INITIAL ).
  variable = 42.
ENDIF.
```

Expected:
``` abap
IF variable IS INITIAL.
  variable = 42.
ENDIF.
```

**Comment by jmgrassau** [comment]
Hi Jan, 

from "thumbs up" response, definitely our most popular proposal so far! :-) 

I completely agree with your example, but I wonder whether in other cases, parentheses might actually be intended to improve readability, e.g. by grouping subexpressions that semantically belong together …

```ABAP
  IF lv_is_valid = abap_true AND ( lv_year = 2023 AND lv_month < 10 ).
  ENDIF.
```

… or because not everyone may be so familiar with operator precedence rules:

```ABAP
  IF ( lv_year = 2022 AND lv_month > 8 ) OR ( lv_year = 2023 AND lv_month < 10 ).
  ENDIF.
```

But that could of course be solved with options, e.g. "Remove parentheses around subexpressions…":
* never
* if all logical operators are the same
* whenever operator precedence allows

Kind regards,
Jörg-Michael

**Comment by ConjuringCoffee** [comment]
Good thinking. I think I'd choose "Never" for the option "Remove parentheses around subexpressions". I'd only be interested in removing the parentheses if the entire condition is wrapped in parentheses.

**Comment by jmgrassau** [comment]
Hi all, 

so here's a preview of the new cleanup rule "**Remove needless parentheses**" for the next release: I'd suggest the following options, of which only the first one would be activated by default:

![image](https://github.com/SAP/abap-cleaner/assets/79639298/b2fd056c-8d35-489a-94e3-e663285fb57e)

The first option, "Remove needless parentheses **around the entire logical expression**", does the following (covering 61% of the 2000+ statements with needless parentheses that I found in sample code):
![image](https://github.com/SAP/abap-cleaner/assets/79639298/bb074bc5-26cb-473b-9554-cedd7dcc2adc)

The second option, "Remove needless parentheses around **relational expressions**" (11% of the cases) still looks rather tame and harmless to me: 
![image](https://github.com/SAP/abap-cleaner/assets/79639298/0c72c247-5985-4dfd-b19f-ef088c825d0e)

Third option, "Remove needless parentheses from **OR ( ... AND ... )**" (26%) would be a bit more daring (see discussion above):
![image](https://github.com/SAP/abap-cleaner/assets/79639298/0f8cfbf1-5b95-422b-8755-30cb1fa8605a)

Fourth option, "Remove needless parentheses from **AND ( ... AND ... ) / OR ( ... OR ... )**" (1%):
![image](https://github.com/SAP/abap-cleaner/assets/79639298/9c36530f-7596-4187-bb5a-ccc93a30f490)

The only cases of needless parentheses that I found in sample code NOT covered by these four options would be OR ( NOT ... ) as well as AND ( NOT ... ), but those were so rare that I didn't think an option would be needed. Also, theoretically, you could remove parentheses from NOT ( rel_expr ), e.g. NOT ( a = b ), but for that, I didn't find a single case.

Hope this makes sense, including the proposed default settings?

@fabianlupa:
>  I cannot for the life of me remember the precedence rules 

Maybe try "**OrANGe**" as a mnemonic (from weakest to strongest precedence):
* **Or**
* **A**nd
* **N**ot
* **Ge** (greater or equal, >=) as a representative of any relational expression (i.e. comparison or predicate)

EQUIV would be even weaker than "OR", but that's so weak and so rarely used, it's not even mentioned in the mnemonic ;-)

Kind regards,
Jörg-Michael

**Comment by jmgrassau** [comment]
… so you could say, as long as you well-structured logical expression appears in **OrANGe** 🍊 order…

```ABAP
CHECK            a  = b
         AND NOT c GE d
      OR         a  = c
         AND NOT c GE b.
```

you're fine without any parentheses :-)

**Comment by fabianlupa** [comment]
Thanks for the programming lesson @jmgrassau , shouldn't be too difficult to remember, coincidentally _orange_ is part of the name of my current employer... 🍊

**Comment by jmgrassau** [comment]
Hi Jan, 

thanks again for opening this issue! The new cleanup rule "Remove needless parentheses" is now available in version 1.11.0 which was just released!

Kind regards,
Jörg-Michael
